tests/runtests.py

```diff
@@ -142,13 +143,26 @@ def get_test_modules(gis_enabled):
                 yield test_module
 
 
+def get_label_module(label):
+    """Return the top-level module part for a test label."""
+    path = Path(label)
+    if len(path.parts) == 1:
+        # Then interpret the label as a dotted module name.
+        return label.split('.')[0]
+
+    # Otherwise, interpret the label as a path.
+    path = path.absolute()
+    rel_path = path.relative_to(RUNTESTS_DIR)
+    return rel_path.parts[0]
```

Comment:
I wonder if we could support running `runtests.py` from different directories :thinking: like we do for dotted module names, e.g.
```bash
~/repo/django> ./tests/runtests.py backends.postgresql
```
works fine, but
```bash
~/repo/django> ./tests/runtests.py backends/postgresql/
....
  File "./tests/runtests.py", line 155, in get_label_module
    rel_path = path.relative_to(RUNTESTS_DIR)
  File "/usr/lib/python3.8/pathlib.py", line 904, in relative_to
    raise ValueError("{!r} does not start with {!r}"
ValueError: '/repo/django/backends/postgresql' does not start with '/repo/django/tests'
```
crashes.

I tried to fix this with:
```python
    # Otherwise, interpret the label as a path.
    if not path.is_absolute():
        return path.parts[0]
    else:
        path = path.absolute()
        rel_path = path.relative_to(RUNTESTS_DIR)
        return rel_path.parts[0]
```
but it crashes with `ModuleNotFoundError` (like without this patch):
```
======================================================================
ERROR: backends/postgresql (unittest.loader._FailedTest)
----------------------------------------------------------------------
ImportError: Failed to import test module: backends/postgresql
Traceback (most recent call last):
  File "/usr/lib/python3.8/unittest/loader.py", line 154, in loadTestsFromName
    module = __import__(module_name)
ModuleNotFoundError: No module named 'backends/postgresql'
```

Reply:
> I wonder if we could support running runtests.py from different directories 🤔 like we do for dotted module names

That case did cross my mind, but I stopped thinking about it because I wasn't sure if it was supported.

In terms of what you tried, shouldn't it be the following since you should be passing a directory relative to your working directory? Does it work then?

```
$ ./tests/runtests.py tests/backends/postgresql/
```

Reply:
The code could do an existence check for the passed directory to fail earlier with a better message.

Reply:
> In terms of what you tried, shouldn't it be the following since you should be passing a directory relative to your working directory? Does it work then?
> 
> ```
> $ ./tests/runtests.py tests/backends/postgresql/
> ```
It also crashes but like in https://github.com/django/django/pull/14180#pullrequestreview-624099857

Reply:
Oh, of course.. I'll give it a thought.

Reply:
Wait, what error message are you getting? I'm not able to reproduce your failure.

Reply:
```bash
~/repo/django> ./tests/runtests.py backends/postgresql
Testing against Django installed in '/home/felixx/repo/django/django' with up to 8 processes
backends
Found 1 test(s).
System check identified no issues (0 silenced).
E
======================================================================
ERROR: backends/postgresql (unittest.loader._FailedTest)
----------------------------------------------------------------------
ImportError: Failed to import test module: backends/postgresql
Traceback (most recent call last):
  File "/usr/lib/python3.8/unittest/loader.py", line 154, in loadTestsFromName
    module = __import__(module_name)
ModuleNotFoundError: No module named 'backends/postgresql'


----------------------------------------------------------------------
Ran 1 test in 0.000s

FAILED (errors=1)

~/repo/django> ./tests/runtests.py tests/backends/postgresql
Testing against Django installed in '/repo/django/django' with up to 8 processes
tests
Found 23 test(s).
Creating test database for alias 'default'...
Cloning test database for alias 'default'...
Cloning test database for alias 'default'...
Cloning test database for alias 'default'...
Creating test database for alias 'other'...
Cloning test database for alias 'other'...
Cloning test database for alias 'other'...
Cloning test database for alias 'other'...
System check identified no issues (0 silenced).
EEE....................
======================================================================
ERROR: backends.postgresql.test_introspection (unittest.loader._FailedTest)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/usr/lib/python3.8/unittest/case.py", line 60, in testPartExecutor
    yield
  File "/usr/lib/python3.8/unittest/case.py", line 676, in run
    self._callTestMethod(testMethod)
  File "/usr/lib/python3.8/unittest/case.py", line 633, in _callTestMethod
    method()
  File "/usr/lib/python3.8/unittest/loader.py", line 34, in testFailure
    raise self._exception
ImportError: Failed to import test module: backends.postgresql.test_introspection
Traceback (most recent call last):
  File "/usr/lib/python3.8/unittest/loader.py", line 436, in _find_test_path
    module = self._get_module_from_name(name)
  File "/usr/lib/python3.8/unittest/loader.py", line 377, in _get_module_from_name
    __import__(name)
  File "/repo/django/tests/backends/postgresql/test_introspection.py", line 6, in <module>
    from ..models import Person
  File "/repo/django/tests/backends/models.py", line 8, in <module>
    class Square(models.Model):
  File "/repo/django/django/db/models/base.py", line 113, in __new__
    raise RuntimeError(
RuntimeError: Model class backends.models.Square doesn't declare an explicit app_label and isn't in an application in INSTALLED_APPS.

======================================================================
ERROR: backends.postgresql.test_operations (unittest.loader._FailedTest)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/usr/lib/python3.8/unittest/case.py", line 60, in testPartExecutor
    yield
  File "/usr/lib/python3.8/unittest/case.py", line 676, in run
    self._callTestMethod(testMethod)
  File "/usr/lib/python3.8/unittest/case.py", line 633, in _callTestMethod
    method()
  File "/usr/lib/python3.8/unittest/loader.py", line 34, in testFailure
    raise self._exception
ImportError: Failed to import test module: backends.postgresql.test_operations
Traceback (most recent call last):
  File "/usr/lib/python3.8/unittest/loader.py", line 436, in _find_test_path
    module = self._get_module_from_name(name)
  File "/usr/lib/python3.8/unittest/loader.py", line 377, in _get_module_from_name
    __import__(name)
  File "/repo/django/tests/backends/postgresql/test_operations.py", line 7, in <module>
    from ..models import Person, Tag
  File "/repo/django/tests/backends/models.py", line 8, in <module>
    class Square(models.Model):
  File "/repo/django/django/db/models/base.py", line 113, in __new__
    raise RuntimeError(
RuntimeError: Model class backends.models.Square doesn't declare an explicit app_label and isn't in an application in INSTALLED_APPS.
...

```

Reply:
Are you sure you ran both commands above in the PR's branch?

When I run the first one in the PR's branch, I don't get `ModuleNotFoundError: No module named 'backends/postgresql'`. But I do get that when running the command from main. When I run the first one in the PR's branch, I get this:

```
./tests/runtests.py backends/postgresql
Testing against Django installed in '/.../django/django'
Traceback (most recent call last):
  File "/.../django/./tests/runtests.py", line 643, in <module>
    failures = django_tests(
  File "/.../django/./tests/runtests.py", line 361, in django_tests
    test_labels, state = setup_run_tests(verbosity, start_at, start_after, test_labels)
  File "/.../django/./tests/runtests.py", line 284, in setup_run_tests
    test_modules, state = setup_collect_tests(start_at, start_after, test_labels=test_labels)
  File "/.../django/./tests/runtests.py", line 253, in setup_collect_tests
    test_modules = list(get_filtered_test_modules(
  File "/.../django/./tests/runtests.py", line 165, in get_filtered_test_modules
    test_module = get_label_module(label)
  File "/.../django/./tests/runtests.py", line 155, in get_label_module
    rel_path = path.relative_to(RUNTESTS_DIR)
  File "/.../python3.9/pathlib.py", line 928, in relative_to
    raise ValueError("{!r} is not in the subpath of {!r}"
ValueError: '/.../django/backends/postgresql' is not in the subpath of '/.../django/tests' OR one path is relative and the other is absolute.
```

Reply:
Or maybe you ran both commands with the changes you made to my PR?

Reply:
I run both with:
```python
    # Otherwise, interpret the label as a path.
    if not path.is_absolute():
        return path.parts[0]
    else:
        path = path.absolute()
        rel_path = path.relative_to(RUNTESTS_DIR)
        return rel_path.parts[0]
```
without it:
```
~/repo/django> ./tests/runtests.py backends/postgresql
```
works and
```
~/repo/django> ./tests/runtests.py backends/postgresql/
....
  File "./tests/runtests.py", line 155, in get_label_module
    rel_path = path.relative_to(RUNTESTS_DIR)
  File "/usr/lib/python3.8/pathlib.py", line 904, in relative_to
    raise ValueError("{!r} does not start with {!r}"
ValueError: '/repo/django/backends/postgresql' does not start with '/repo/django/tests'
```
crashes. Sorry for confusing.

Reply:
I was [originally asking](https://github.com/django/django/pull/14507#discussion_r648185318) what error you're getting when you run the following from my PR (with the correct path):

```
$ ./tests/runtests.py tests/backends/postgresql/
```

I'm able to run all of the following with no errors from my PR:

```
./tests/runtests.py tests/view_tests
./tests/runtests.py tests/view_tests/
./tests/runtests.py tests/view_tests/tests/
./tests/runtests.py tests/backends/
./tests/runtests.py tests/backends/sqlite/
./tests/runtests.py tests/backends/postgresql/
```

Reply:
Yeah it works for me, sorry again. The current version looks good :+1: , we could only raise a more descriptive error when a relative path is not correct (as proposed in https://github.com/django/django/pull/14507#discussion_r648186310).

Reply:
Okay, thanks! Will make that update!

Reply:
Updated.